great_ape_safe/ape_api/snapshot.py

```diff
@@ -51,34 +57,65 @@ def _get_proposal_data(self, proposal_id):
             json={"query": self.proposal_query, "variables": {"proposal_id": proposal_id}},
         )
 
-        if not response.ok:
-            print(f"Error getting proposal data: {response.text}")
-            raise
-
+        self.handle_response(response)
         return response.json()["data"]["proposal"]
 
 
     def show_proposal_choices(self):
         # external helper method to view proposal choices
         choices = self.proposal_data["choices"]
         console.print(f"Choices for proposal {self.proposal_id}: {choices}")
-    
+
+
+    def create_payload_hash(
+            self,
+            payload=None, 
+            timestamp=None, 
+            proposal=None, 
+            choice=None, 
+            version="0.1.3", 
+            type="vote",
+            metadata=""
+        ):
+        # helper method to create message hash from payload and output to console
+        # can be used externally to verify generated hash
+        if not payload:
+            assert all([timestamp, proposal, choice])
+            payload = {
+                "version": version,
+                "timestamp": timestamp,
+                "space": self.proposal_data["space"]["id"],
+                "type": type,
+                "payload": {
+                    "proposal": proposal,
+                    "choice": choice,
```

Comment:
i think in the choice depends in which format is passed may need a bit more of logic.

there'd be couple of scenarios, but to easen up the review process i think in the choice should be pass as a string the choice and then in the script check if it's a weighted proposal or not, then accordingly should do similar logic like in this [section](https://github.com/Badger-Finance/badger-multisig/blob/main/great_ape_safe/ape_api/snapshot.py#L79-L88). otherwise the `choice` may not have the right format when recreating the payload

Reply:
this is looking good for the current for/against proposal that is open:
```
>>> from great_ape_safe import GreatApeSafe
>>> safe=GreatApeSafe('techops.badgerdao.eth')
>>> safe.init_snapshot('0xc8c920dc682c4c56ad4d4ed40c328d0022d8badf903bc6ed7e7fa345c9aab5c7')
>>> safe.snapshot.vote_and_post('For')
payload
{
    'version': '0.1.3',
    'timestamp': '1660732163',
    'space': 'aurafinance.eth',
    'type': 'vote',
    'payload': {'proposal': '0xc8c920dc682c4c56ad4d4ed40c328d0022d8badf903bc6ed7e7fa345c9aab5c7', 'choice': '1', 'metadata': ''}
}
msg hash: 0x756ffecf4e46d7067bd955362e3fa27bde251fc2cdc5894bdf0bd83986a10e75
signing vote for choice 1
{
│   'ethereum_client': <gnosis.eth.ethereum_client.EthereumClient object at 0x1133bc220>,
│   'safe_address': '0x86cbD0ce0c087b482782c181dA8d191De18C8275',
│   'to': '0xA65387F16B013cf2Af4605Ad8aA5ec25a2cbA3a2',
│   'value': 0,
│   'data': HexBytes('0x85a5affe00000000000000000000000000000000000000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000020756ffecf4e46d7067bd955362e3fa27bde251fc2cdc5894bdf0bd83986a10e75'),
│   'operation': 1,
│   'safe_tx_gas': 0,
│   'base_gas': 0,
│   'gas_price': 0,
│   'gas_token': '0x0000000000000000000000000000000000000000',
│   'refund_receiver': '0x0000000000000000000000000000000000000000',
│   'signatures': b'',
│   '_safe_nonce': 336,
│   '_safe_version': '1.1.1',
│   '_chain_id': None
}
```

lets give it another shot with a weighted prop later.

Reply:
i still think that the choice needs to catch if there is a weighted vote type and stringify similar like [here](https://github.com/Badger-Finance/badger-multisig/blob/issue/725/great_ape_safe/ape_api/snapshot.py#L116-L120)

Reply:
can you confirm @gosuto-inzasheru  @jalbrekt85 you're getting the same hash with the create payload?. locally for me does not match running `brownie run snapshot/AURA/vote_proposal` and passing the choice into the method as raw, i believe for the cause above i mention or i may be missing something:

```
payload
{
    'version': '0.1.3',
    'timestamp': '1660830370',
    'space': 'aurafinance.eth',
    'type': 'vote',
    'payload': {
        'proposal': '0x632726577e9c186a8827311b8702dcfd98f26376237b9715e64eafa61d6d7672',
        'choice': '{"32":1}',
        'metadata': ''
    }
}
msg hash: 0xe5e20333e078bbb430e6a8aac80e8c6cc880b3e378379076f22f6b02de070287
```

VS

```
>>> voter.snapshot.create_payload_hash(timestamp=1660830370, proposal='0x632726577e9c186a8827311b8702dcfd98f26376237b9715e64eafa61d
6d7672', choice='{"32":1}')
msg hash: 0x51ccfdd91225276ded262350f95ecf60780872b848abdde34a13c4e2d08c2092
(b"Q\xcc\xfd\xd9\x12%'m\xed&#P\xf9^\xcf`x\x08r\xb8H\xab\xdd\xe3J\x13\xc4\xe2\xd0\x8c \x92", "{"version":"0.1.3","timestamp":1660830370,"space":"aurafinance.eth","type":"vote","payload":{"proposal":"0x632726577e9c186a8827311b8702dcfd98f26376237b9715e64eafa61d6d7672","choice":"{\"32\":1}","metadata":""}}")
```

passing the payload outputted via `brownie run snapshot/AURA/vote_proposal` will generate using `create_payload_hash` same hash

Reply:
looks like the discrepancy between outputted hashes from these methods comes from the `timestamp` not being casted to str in `create_payload_hash`

Reply:
@petrovska-petro addressed your comments in [9d85c83](https://github.com/Badger-Finance/badger-multisig/pull/737/commits/9d85c83b010fb437b0e8be55329d8936e8fb99b2). `create_payload_hash` will get the choice index for single/weighted from "raw" choice. payload hashes should match now

Reply:
very sweet and great refactoring! all matching now 🔥